src/jit/decompose.cpp

```diff
@@ -0,0 +1,1189 @@
+        {
+            insertPt = insertPt->gtNext;
+        }
+
+        comp->fgInsertStmtAfter(comp->compCurBB, insertPt, newStmt);
+    }
+    else
+    {
+        // The current statement is an embedded statement. Create a new embedded statement to
+        // contain the node. First, find the parent non-embedded statement containing the
+        // current statement.
+        GenTree* parentStmt = curStmt;
+        while ((parentStmt != nullptr) && (!parentStmt->AsStmt()->gtStmtIsTopLevel()))
+        {
+            parentStmt = parentStmt->gtPrev;
+        }
+        assert(parentStmt != nullptr);
+
+        newStmt = comp->fgMakeEmbeddedStmt(comp->compCurBB, node, parentStmt);
+    }
+
+    newStmt->gtStmtILoffsx = curStmt->gtStmtILoffsx;
+#ifdef DEBUG
+    newStmt->gtStmtLastILoffs = curStmt->gtStmtLastILoffs;
+#endif // DEBUG
+}
+
+
+//------------------------------------------------------------------------
+// GetHiOper: Convert arithmetic operator to "high half" operator of decomposed node.
+//
+// Arguments:
+//    oper - operator to map
+//
+// Return Value:
+//    mapped operator
+//
+genTreeOps Decompose::GetHiOper(genTreeOps oper)
+{
+    switch(oper)
+    {
+    case GT_ADD: return GT_ADD_HI;  break;
+    case GT_SUB: return GT_SUB_HI;  break;
+    case GT_MUL: return GT_MUL_HI;  break;
+    case GT_DIV: return GT_DIV_HI;  break;
+    case GT_MOD: return GT_MOD_HI;  break;
+    case GT_OR:  return GT_OR;      break;
+    case GT_AND: return GT_AND;     break;
+    case GT_XOR: return GT_XOR;     break;
+    default:
+        assert(!"GetHiOper called for invalid oper");
+        return GT_NONE;
+    }
+}
+
+
+//------------------------------------------------------------------------
+// GetLoOper: Convert arithmetic operator to "low half" operator of decomposed node.
+//
+// Arguments:
+//    oper - operator to map
+//
+// Return Value:
+//    mapped operator
+//
+genTreeOps Decompose::GetLoOper(genTreeOps oper)
+{
+    switch(oper)
+    {
+    case GT_ADD: return GT_ADD_LO;  break;
+    case GT_SUB: return GT_SUB_LO;  break;
+    case GT_OR:  return GT_OR;      break;
+    case GT_AND: return GT_AND;     break;
+    case GT_XOR: return GT_XOR;     break;
+    default:
+        assert(!"GetLoOper called for invalid oper");
+        return GT_NONE;
+    }
+}
+
+
+//------------------------------------------------------------------------
+// SimpleLinkNodeAfter: insert a node after a given node in the execution order.
+// NOTE: Does not support inserting after the last node of a statement, which
+// would require updating the statement links.
+//
+// Arguments:
+//    insertionPoint - Insert after this tree node.
+//    node - The node to insert.
+//
+// Return Value:
+//    None
+//
+// Notes:
+// Seems like this should be moved to someplace that houses all the flowgraph
+// manipulation functions.
+//
+void Decompose::SimpleLinkNodeAfter(GenTree* insertionPoint, GenTree* node)
```

Comment:
Completely agree. I didn't really feel like changing it now, though. I've been thinking it would be good to move all the flowgraph utility manipulation function to a common location, regularize their names, and make sure it is clear which work pre- and post-lower, and so forth.